Rewrite methods to use async calls of HttpClient and Uni

skyllarr opened the pull request.

skyllarr committed:
Add AsyncAcmeClientSpi class whose methods return Uni values

skyllarr committed:
Add tests for AsyncClientSpi class

skyllarr committed:
Rewrite methods to use asynchronous calls with HttpClient and Mutiny

sberyozkin commented:
@cescoffier Please review this PR, thanks

cescoffier commented on x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java:
```diff
@@ -0,0 +1,282 @@
+        if (account == null) {
+            return Uni.createFrom().failure(new IllegalArgumentException("account"));
+        }
+        HttpClient httpClient = HttpClient.newHttpClient();
```
Unrelated to this PR but Quarkus is avoiding using the JDK client.
Should we consider adding an SPI?

sberyozkin replied:
Quarkus Keycloak Authorization uses Keycloak authorization HttpRequest/HttpResponse SPI, see https://github.com/quarkusio/quarkus/blob/main/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/VertxHttpFacade.java, may be something similar can be considered

skyllarr replied:
@cescoffier @sberyozkin  Thank you, I have created HttpRequestSpi, HttpResponseSpi, and HttpClientSpi so that any other client could be used. You can see how they look [here](https://github.com/Skyllarr/certificate-management/tree/async-spi-with-rewrite/x500/cert/spi/src/main/java/org/wildfly/security/certificate/management/x500/cert/spi).

In my tests I used standard JDK client for now. Which HTTP client would you prefer? I can use the preferred one in my tests instead

cescoffier replied:
We will use the Vert.x Web client exposing the Mutiny API: https://github.com/smallrye/smallrye-mutiny-vertx-bindings/blob/main/vertx-mutiny-clients/vertx-mutiny-web-client/src/test/java/io/vertx/mutiny/web/WebClientTest.java#L41

cescoffier commented on x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java:
```diff
@@ -0,0 +1,282 @@
+                            .uri(newNonceUri)
+                            .header(ACCEPT_LANGUAGE, Locale.getDefault().toLanguageTag())
+                            .header(USER_AGENT, USER_AGENT_STRING).build();
+                    return Uni.createFrom().future(httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString()));
```
That's exactly why we do not want to use this HTTP client. 

sendAsync is not emitting on an uncontrolled F/J thread. 
Also, creating a Uni from a future may be blocking (if not an instance of CompletionStage, which in this case is not a problem).

skyllarr replied:
@cescoffier  The SPI I added also returns Future, see [here](https://github.com/Skyllarr/certificate-management/blob/async-spi-with-rewrite/x500/cert/spi/src/main/java/org/wildfly/security/certificate/management/x500/cert/spi/HttpClientSpi.java#L24), which is then used to create Uni. Is this okay or did you have a different approach in mind? 

I can change the return value to be CompleteableFuture instead, since the CompleteableFuture implements the CompletionStage, if that is better?

Thank you for your help!

cescoffier replied:
`Future` are _blocking_, in the sense you cannot listen for the outcome without blocking. So, it's not great.

`CompletionStage` is the best you can do (or returns a `Uni`). At least you can listen for the result without blocking it, and we can create instances of Uni from `CompletionStage`. 

However, make sure you never use the `*Async` method from the CompletionStage API.

sberyozkin replied:
@Skyllarr @cescoffier Should `Uni` be indeed introduced right now in the client SPI, since `Uni` is already featuring in other parts of this project ?

skyllarr replied:
@cescoffier Sounds good, I have updated the `HttpClientSpi` to return the `CompletableFuture` instead of the `Future`, since `CompletableFuture` implements `CompletionStage`. Thanks for explaining!

@sberyozkin `Uni` instances are being returned in the `AsyncAcmeClientSpi`. This is so that this SPI is not blocking and we can handle the failures with `onFailure()` there for example. Which other parts of the project you had in mind that would introduce `Uni` instead of the `AsyncAcmeClientSpi`? It is this SPI that should be used in Quarkus code

cescoffier commented on x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java:
```diff
@@ -0,0 +1,282 @@
+    }
+
+    private Uni<HttpResponse<String>> sendGetRequest(String resourceUrl, int expectedResponseCode, String expectedContentType) {
+        HttpClient httpClient = HttpClient.newBuilder().build();
```
Are we creating an instance of HTTP client per call?

skyllarr replied:
Good point, updated to not create new instance each time

cescoffier commented on x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java:
```diff
@@ -0,0 +1,282 @@
+            httpRequest = HttpRequest.newBuilder().GET().uri(new URI(resourceUrl))
+                    .header(ACCEPT_LANGUAGE, Locale.getDefault().toLanguageTag())
+                    .header(USER_AGENT, USER_AGENT_STRING).build();
+            return Uni.createFrom().future(httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString()))
```
Same comment as above.

skyllarr replied:
Updated

cescoffier commented on x500/cert/acme/src/main/java/org/wildfly/security/certificate/management/x500/cert/acme/AsyncAcmeClientSpi.java:
```diff
@@ -0,0 +1,282 @@
+                    .header(ACCEPT_LANGUAGE, Locale.getDefault().toLanguageTag())
+                    .header(USER_AGENT, USER_AGENT_STRING).build();
+            return Uni.createFrom().future(httpClient.sendAsync(httpRequest, HttpResponse.BodyHandlers.ofString()))
+                    .onItem().call((HttpResponse<String> httpResponse) -> {
```
can be a `chain` with the current implementation.

skyllarr replied:
Updated, thank you!

skyllarr committed:
Add SPIs for HttpRequest, HttpResponse and HttpClient

skyllarr committed:
Minor refactor

skyllarr committed:
Change future to CompletableFuture in HttpClientSpi and add getMetadata method

skyllarr committed:
WIP Rewrite more methods to use Uni, this needs refactoring and rewrite of the remaining methods

skyllarr commented:
I added a new commit with more methods rewritten, this needs some refactoring and mostly rewriting of obtainCertificateChain method. Hopefully I'll be able submit this on Monday.

Some thoughts though: I was rewriting this API by not changing the method headers so that it mirrors the old synchronous API. But this is probably not right, because there is an object `account` that is being passed to and modified by some methods. In the async environment, this could introduce concurrency issues. It would be best have an immutable input, deep copy the account object and Uni should deliver the result. This is also a TODO task

sberyozkin commented:
Thanks @Skyllarr
